test/ex_plasma/transaction_test.exs

```diff
@@ -1,328 +1,432 @@
 
+  alias ExPlasma.Builder
+  alias ExPlasma.Output
+  alias ExPlasma.Support.TestEntity
   alias ExPlasma.Transaction
+  alias ExPlasma.Transaction.Type.PaymentV1
+  alias ExPlasma.Transaction.TypeMapper
+
+  @alice TestEntity.alice()
+  @bob TestEntity.bob()
+  @eth <<0::160>>
+  @zero_metadata <<0::256>>
+  @payment_tx_type TypeMapper.tx_type_for(:tx_payment_v1)
+  @payment_output_type TypeMapper.output_type_for(:output_payment_v1)
+
+  setup_all do
+    %{priv_encoded: alice_priv, addr: alice_addr} = @alice
+    %{addr: bob_addr} = @bob
+
+    signed =
+      ExPlasma.payment_v1()
+      |> Builder.new()
+      |> Builder.add_input(blknum: 1, txindex: 0, oindex: 0, position: 1_000_000_000)
+      |> Builder.add_input(blknum: 2, txindex: 0, oindex: 0, position: 2_000_000_000)
+      |> Builder.add_output(output_guard: bob_addr, token: @eth, amount: 12)
+      |> Builder.sign!([alice_priv, alice_priv])
+
+    encoded_signed_tx = Transaction.encode(signed)
+
+    [sigs, _payment_marker, inputs, outputs, _txdata, _metadata] = ExRLP.decode(encoded_signed_tx)
+
+    {:ok,
+     %{
+       alice_addr: alice_addr,
+       bob_addr: bob_addr,
+       signed: signed,
+       encoded_signed_tx: encoded_signed_tx,
+       sigs: sigs,
+       inputs: inputs,
+       outputs: outputs
+     }}
+  end
 
-  @transaction %Transaction{
-    inputs: [
-      %ExPlasma.Output{
-        output_data: nil,
-        output_id: %{blknum: 0, oindex: 0, position: 0, txindex: 0},
-        output_type: nil
-      }
-    ],
-    metadata: <<0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0>>,
-    outputs: [
-      %ExPlasma.Output{
-        output_data: %{
-          amount: 10,
-          output_guard: <<21, 248, 47, 41, 27, 46, 150, 159, 176, 132, 157, 153, 217, 206, 65, 226, 241, 55, 0, 110>>,
-          token: <<46, 38, 45, 41, 28, 46, 150, 159, 176, 132, 157, 153, 217, 206, 65, 226, 241, 55, 0, 110>>
-        },
-        output_id: nil,
-        output_type: 1
-      }
-    ],
-    sigs: [],
-    tx_data: <<0>>,
-    tx_type: 1
-  }
-
-  describe "encode/1" do
-    test "encodes a transaction struct" do
-      result = Transaction.encode(@transaction)
+  describe "decode/2" do
+    test "decodes successfuly in various empty input/output combinations" do
+      transaction_list = [
+        {[], [], [{@alice, @eth, 7}]},
+        {[{1, 2, 3}], [@alice], [{@alice, @eth, 7}]},
+        {[{1, 2, 3}], [@alice], [{@alice, @eth, 7}, {@bob, @eth, 3}]},
+        {[{1, 2, 3}, {2, 3, 4}], [@alice, @bob], [{@alice, @eth, 7}, {@bob, @eth, 3}]},
+        {[{1, 2, 3}, {2, 3, 4}, {2, 3, 5}], [@alice, @bob, @bob], [{@alice, @eth, 7}, {@bob, @eth, 3}]},
+        {[{1, 2, 3}, {2, 3, 4}, {2, 3, 5}], [@alice, @bob, @bob],
+         [{@alice, @eth, 7}, {@bob, @eth, 3}, {@bob, @eth, 3}]},
+        {[{1, 2, 3}, {2, 3, 1}, {2, 3, 2}, {3, 3, 4}], [@alice, @alice, @bob, @bob],
+         [{@alice, @eth, 7}, {@alice, @eth, 3}, {@bob, @eth, 7}, {@bob, @eth, 3}]}
+      ]
```

Comment:
eh nitpick, not a fan of this way to enumerate over multiple transaction combinations, but I also don't have a strong recommendation currently.